pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/NativeTextIndexCreator.java

```diff
@@ -0,0 +1,173 @@
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import org.apache.commons.io.FileUtils;
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.standard.StandardAnalyzer;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.pinot.segment.local.segment.creator.impl.inv.BitmapInvertedIndexWriter;
+import org.apache.pinot.segment.local.segment.creator.impl.text.LuceneTextIndexCreator;
+import org.apache.pinot.segment.local.utils.nativefst.builder.FSTBuilder;
+import org.apache.pinot.segment.spi.V1Constants;
+import org.apache.pinot.segment.spi.index.creator.TextIndexCreator;
+import org.roaringbitmap.Container;
+import org.roaringbitmap.RoaringBitmap;
+import org.roaringbitmap.RoaringBitmapWriter;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+
+public class NativeTextIndexCreator implements TextIndexCreator {
+    static final String TEMP_DIR_SUFFIX = ".nativetext.idx.tmp";
+    static final String FST_FILE_NAME = "native.fst";
+    static final String INVERTED_INDEX_FILE_NAME = "inverted.index.buf";
+    public static final int HEADER_LENGTH = 24;
+
+    private String _columnName;
+    private FSTBuilder _fstBuilder;
+    private final File _indexFile;
+    private final File _tempDir;
+    private final File _fstIndexFile;
+    private final File _invertedIndexFile;
+    final Map<String, RoaringBitmapWriter<RoaringBitmap>> _postingListMap = new TreeMap<>();
+    final RoaringBitmapWriter.Wizard<Container, RoaringBitmap> _bitmapWriterWizard = RoaringBitmapWriter.writer();
+    private int _nextDocId = 0;
+    private int _fstDataSize;
+    private int _numBitMaps;
+
+    public NativeTextIndexCreator(String column, File indexDir)
+            throws IOException {
+        _columnName = column;
+        _fstBuilder = new FSTBuilder();
+        _indexFile = new File(indexDir, column + V1Constants.Indexes.NATIVE_TEXT_INDEX_FILE_EXTENSION);
+        _tempDir = new File(indexDir, column + TEMP_DIR_SUFFIX);
+        if (_tempDir.exists()) {
+            FileUtils.cleanDirectory(_tempDir);
+        } else {
+            FileUtils.forceMkdir(_tempDir);
+        }
+        _fstIndexFile = new File(_tempDir, FST_FILE_NAME);
+        _invertedIndexFile = new File(_tempDir, INVERTED_INDEX_FILE_NAME);
+    }
+
+    @Override
+    public void add(String document) {
+        List<String> tokens;
+        try {
+            tokens = analyze(document, new StandardAnalyzer(LuceneTextIndexCreator.ENGLISH_STOP_WORDS_SET));
+        } catch (IOException e) {
+            throw new RuntimeException(e.getMessage());
+        }
+
+        for (String token : tokens) {
+            addToPostingList(token);
+        }
+
+        _nextDocId++;
+    }
+
+    @Override
+    public void add(String[] documents, int length) {
+        for (int i = 0; i < length; i++) {
+            add(documents[i]);
+        }
```

Comment:
This looks wrong, the API is for MV columns, but each of the values would have a different docId implemented this way, and the docId will get out of sync with the docId used for indexes on other attributes of the same record and for the storage. This will lead to nonsensical query results when filtering on MV text attributes.

Without a secondary docId (in the JSON index this is called the flattened docId) I don't see how MV columns could work without cross-match after tokenisation. The Lucene text index also looks wrong for MV columns in this respect. Do we have some testing for a query like

```sql
select someOtherAttribute from table where mvTextIndexedAttribute like '%xyz%`
```

with records:

| someOtherAttribute | mvTextIndexedAttribute |
|---------------------|-------------------------|
| 0                                |. ['abc', 'xyz'].                  |
| 1                                |. ['def', 'uvw'].                  |

`like 'xyz'` will produce the bitmap {1}, which will then be used to access the wrong value from the forward index for `someOtherAttribute`

Reply:
To clarify my expectations on how text matches on MV columns should work: I would expect a regular expression matching _any_ term in an MV array to select the document, but I would not expect cross-match within the MV array. So in the example above, "a.*c" or "xy.*" would match record 0, but a regular expression which would match a conflation of the terms (e.g. "a.*z") but not any of the individual terms must not.

Reply:
I am trying to parse this comment. Is the concern the fact that the multi add document API is assigning different docIDs per document? That is easy to fix, but I am confused. The API documentation in TextIndexCreator says:

`  * Adds a set of documents to the index`

This does not talk about treating all documents as the same field, nor about this being the MV API. What am I missing?

Reply:
tl;dr: this method causes multiple docIds to be used for the same doc, which will lead to nonsense results.

Reply:
Per the MV semantic, we should associate all the entries to the same docId, but avoid cross matching them. We should have a test for it and ensure the expected behavior

Reply:
If you look at where it's used (`SegmentColumnarIndexCreator`, `TextIndexHandler`) the expectations will probably be clearer than reading the Javadoc - it's always used for MV columns.

Reply:
The obvious solution here is to use the same docId for each value in the array, but another challenge with MV columns is what happens after tokenisation, which would lead to cross match. Imagine you have an array ["Johnnie Walker Black Label", "Jack Daniels"]  which is tokenised into something like ["johnnie", "walker", "black", "label"] and ["jack", "daniels"], it needs to be ensured that queries for "jack black" don't match with the record. This isn't possible to implement without structural changes to the inverted index because the membership of each token to the term in the MV array is lost. The simplest solution would be to just not support MV columns at present.

Reply:
+1 on not support MV at present. We may add another format when adding the MV support in the future

Reply:
MV support has been removed.

Reply:
Thanks for clarifying. We should *explicitly* mention that the multi add API is for MV columns only. There is nothing in the interface or the documentation that leads to this conclusion, and nor do we ask people to refer to examples for understanding what a mis-named API actually does.

Regarding cross queries, what has been described here is known as a phrase query. In the current implementation, Native text indices will support *only* regex matches using TEXT_MATCH. To support phrase queries, we need an inverted index which has positions. That will be added as a part of next phase.

Removed MV support in this PR.